File: src/backend/app/services/routing.py
"""Alternate routing / diversion recommender with explicit live-data gating.

In-port alternatives are driven by the application's live forecast/optimiser state.
External-port diversion is only eligible when a current port-status feed is supplied
through PORT_STATUS_JSON; the static reference table is used only for physical
feasibility and published reference metadata, never for pretending current berth
availability or congestion is live.
"""

from __future__ import annotations

import json
import os

from .. import reference as ref

DIVERT_WAIT_H = 48.0
SLOW_STEAM_WAIT_H = 18.0
PRIORITY_WAIT_H = 10.0
REEFFER_PRIORITY_MIN = 200
SUSTAINED_HOURS = 3


def _live_port_status() -> dict[str, dict]:
    """Read an operator-provided current external-port feed.

    Format: {"Port of Oakland": {"availability": "medium", "wait_hours": 18,
    "updated_at": "..."}}. Missing/malformed feeds return {} so no external
    diversion is recommended from stale constants.
    """
    raw = os.environ.get("PORT_STATUS_JSON", "").strip()
    if not raw:
        return {}
    try:
        data = json.loads(raw)
        if not isinstance(data, dict):
            return {}
        out = {}
        for name, value in data.items():
            if not isinstance(value, dict):
                continue
            availability = str(value.get("availability", "")).lower()
            wait = float(value.get("wait_hours"))
            if availability not in {"high", "medium", "low"} or wait < 0:
                continue
            out[str(name)] = {**value, "availability": availability, "wait_hours": wait}
        return out
    except (TypeError, ValueError, json.JSONDecodeError):
        return {}


def _predicted_wait(ctx, forecasts, assignment_by_vessel, v) -> tuple[float, bool]:
    a = assignment_by_vessel.get(v.id)
    zone = a["zone_code"] if a else v.dest_zone_code
    fc = forecasts.get(zone) or forecasts.get("Z-PORT")
    if fc is None:
        return v.anchored_hours, False
    start = int(max(1, min(72, round(a["start_hour"] if a else max(1, v.eta_hours)))))
    wait = fc.points[start - 1].wait
    segment = fc.points[start - 1:start - 1 + SUSTAINED_HOURS]
    sustained = len(segment) == SUSTAINED_HOURS and all(p.wait >= DIVERT_WAIT_H for p in segment)
    if a is None:
        return max(wait, 78.0), True
    return wait, sustained


def _best_window(fc) -> tuple[int, float] | None:
    if not fc or not fc.points:
        return None
    p = min(fc.points, key=lambda x: x.wait)
    return p.hour, p.wait


def _best_in_port_alternative(ctx, forecasts: dict, v, current_wait: float, min_saving_h: float = 12.0):
    best = None
    for zone, fc in forecasts.items():
        if zone == "Z-PORT" or zone == v.dest_zone_code:
            continue
        fits = [b for b in ctx.berths if b.zone_code == zone and v.loa_ft <= b.length_ft and v.draft_ft <= b.depth_ft]
        if not fits:
            continue
        w = _best_window(fc)
        if not w:
            continue
        wait_alt = w[1]
        if current_wait - wait_alt >= min_saving_h and (best is None or wait_alt < best[1]):
            best = (zone, wait_alt, round(current_wait - wait_alt, 1))
    return best


def _option_detail(ctx, forecasts: dict, v, wait: float, rule: str) -> dict:
    fc = forecasts.get(v.dest_zone_code) or forecasts.get("Z-PORT")
    window = _best_window(fc)
    alt = _best_in_port_alternative(ctx, forecasts, v, wait)
    return {
        "rule": rule,
        "berthing_window": ({"hour": window[0], "wait_hours": round(window[1], 1)} if window else None),
        "alternate_terminal": ({"zone_code": alt[0], "wait_hours": round(alt[1], 1), "savings_hours": alt[2]} if alt else None),
    }


def recommend_routing(ctx, forecasts: dict, optimiser_out: dict | None) -> list[dict]:
    assignment_by_vessel = {a["vessel_id"]: a for a in (optimiser_out or {}).get("assignments", [])}
    live_ports = _live_port_status()
    recs: list[dict] = []

    for v in ctx.vessels:
        wait, sustained = _predicted_wait(ctx, forecasts, assignment_by_vessel, v)
        base = {"vessel_id": v.id, "vessel_name": v.name, "carrier": v.carrier,
                "vessel_class": v.vessel_class, "dest_zone_code": v.dest_zone_code,
                "status": v.status, "predicted_wait_hours": round(wait, 1)}

        def _detail(rule: str) -> dict:
            return _option_detail(ctx, forecasts, v, wait, rule)

        if wait >= DIVERT_WAIT_H and sustained:
            candidates = []
            for p in ref.ALT_PORTS:
                live = live_ports.get(p["name"])
                if not live:
                    continue
                if v.loa_ft > p["max_loa_ft"] or v.draft_ft > p.get("max_berth_depth_ft", 55.0):
                    continue
                candidates.append((p, live))
            best = None
            for p, live in candidates:
                live_wait = float(live["wait_hours"])
                transit = float(p["transit_hours"])
                shift = transit + live_wait
                wait_avoided = max(0.0, wait - shift)
                savings = ((wait_avoided / 24) * ref.DAILY_OP_COST_USD
                            - (shift / 24) * ref.DAILY_OP_COST_USD * 0.35)
                if best is None or savings > best[1]:
                    best = (p, live, savings, shift, wait_avoided)
            if best and best[2] > 0:
                p, live, savings, shift, wait_avoided = best
                recs.append({**base, "option": "DIVERT", "target_port": p["name"],
                             "eta_shift_hours": round(shift, 1), "est_savings_usd": round(savings),
                             "confidence": 0.9, "tier": "critical", "sustained": True,
                             "option_detail": _detail("divert: sustained forecast + live external port status"),
                             "rationale": (f"Predicted {wait:.0f}h wait; live feed reports {live['wait_hours']:.0f}h at {p['name']} "
                                           f"with {live['availability']} availability. Estimated net savings ≈ ${round(savings):,}.")})
                continue

        if SLOW_STEAM_WAIT_H <= wait < DIVERT_WAIT_H and v.status == "INBOUND":
            steam_down = min(wait - 6, 48)
            savings = 0.35 * (steam_down / 24) * ref.DAILY_OP_COST_USD
            recs.append({**base, "option": "SLOW_STEAM", "target_port": None, "eta_shift_hours": round(steam_down, 1),
                         "est_savings_usd": round(savings), "confidence": 0.72, "tier": "high", "sustained": sustained,
                         "option_detail": _detail("slow-steam: 18h<=wait<48h and INBOUND"),
                         "rationale": f"Inbound with predicted {wait:.0f}h queue; slow-steam to target the forecast opening window. Estimated operating saving ≈ ${round(savings):,}."})
            continue

        if wait >= PRIORITY_WAIT_H and v.reefer_units >= REEFFER_PRIORITY_MIN:
            savings = v.reefer_units * ref.REEFER_CONTENT_VALUE_USD * 0.4
            recs.append({**base, "option": "PRIORITY_WINDOW", "target_port": None,
                         "eta_shift_hours": -min(6, wait - 8), "est_savings_usd": round(savings),
                         "confidence": 0.64, "tier": "medium", "sustained": sustained,
                         "option_detail": _detail("priority window: wait>=10h and >=200 reefers"),
                         "rationale": f"{v.reefer_units} reefer units with {wait:.0f}h predicted wait; prioritise the vessel to reduce spoilage exposure."})
            continue

        recs.append({**base, "option": "HOLD", "target_port": None, "eta_shift_hours": 0, "est_savings_usd": 0,
                     "confidence": 0.55, "tier": "low", "sustained": sustained,
                     "option_detail": _detail("hold: within normal rotation or no validated external alternative"),
                     "rationale": (f"Predicted {wait:.0f}h wait within normal rotation — hold current schedule."
                                   if wait < PRIORITY_WAIT_H else
                                   f"Predicted {wait:.0f}h wait but no validated live external alternative; hold and re-evaluate next cycle.")})

    tier_rank = {"critical": 0, "high": 1, "medium": 2, "low": 3}
    recs.sort(key=lambda r: (tier_rank[r["tier"]], -r["est_savings_usd"]))
    return recs

File: src/lib/engine/routing.ts
// ============================================================================
// Alternate routing recommendation engine (rule-based, documented cost model).
//
// Decides per vessel between: DIVERT / SLOW_STEAM / PRIORITY_WINDOW / HOLD,
// using the congestion forecast at the vessel's destination zone + published
// operating-cost assumptions (all constants cited in docs/solution-overview.md).
// ============================================================================
import type { RoutingRec, VesselInfo } from "./types";

// Documented assumption: daily operating cost (fuel + ops) for a mid/large
// container ship. Public industry estimates range ~$25k–$45k/day; we use $32k.
const DAILY_OP_COST_USD = 32_000;
const REEFER_CONTENT_VALUE_USD = 180; // expected spoilage-risk per reefer unit per event

interface AltPort {
  name: string;
  distanceNm: number;
  transitHrsAt15kn: number;
  // rule-based availability for redirected capacity (demo constant, documented)
  availability: "high" | "medium" | "low";
  maxLoaFt: number;
}

const ALT_PORTS: AltPort[] = [
  { name: "Port of Oakland", distanceNm: 500, transitHrsAt15kn: 33, availability: "medium", maxLoaFt: 1320 },
  { name: "Seattle-Tacoma (NW Seaport)", distanceNm: 1180, transitHrsAt15kn: 79, availability: "low", maxLoaFt: 1320 },
  { name: "Prince Rupert (Fairview)", distanceNm: 1260, transitHrsAt15kn: 84, availability: "low", maxLoaFt: 1300 },
  { name: "Ensenada (ECT)", distanceNm: 150, transitHrsAt15kn: 10, availability: "high", maxLoaFt: 1000 },
];

const availabilityBufferHrs: Record<AltPort["availability"], number> = {
  high: 6,
  medium: 18,
  low: 36,
};

export interface RoutingInput {
  vessels: VesselInfo[];
  // predicted avg anchorage wait (hours) at the vessel's destination zone
  // over the window in which the vessel would be worked
  predictedWaitFor: (v: VesselInfo) => number;
}

export function recommendRouting(input: RoutingInput): RoutingRec[] {
  const recs: RoutingRec[] = [];

  for (const v of input.vessels) {
    const wait = input.predictedWaitFor(v);
    const base = {
      vesselId: v.id,
      vesselName: v.name,
      carrier: v.carrier,
      vesselClass: v.vesselClass,
      destZoneCode: v.destZoneCode,
      status: v.status,
      predictedWaitHrs: +wait.toFixed(1),
    };

    // ---------------------------------------------------------------- DIVERT
    if (wait >= 48) {
      const candidates = ALT_PORTS.filter(
        (p) => v.loaFt <= p.maxLoaFt && p.availability !== "low",
      );
      let best: { port: AltPort; savings: number; shift: number } | null = null;
      for (const p of candidates) {
        // divert cost = extra transit beyond the wait we avoid + terminal buffer
        const shift = Math.max(0, p.transitHrsAt15kn + availabilityBufferHrs[p.availability] - wait);
        const waitAvoided = Math.max(0, wait - shift);
        const savings = (waitAvoided / 24) * DAILY_OP_COST_USD - (shift / 24) * DAILY_OP_COST_USD * 0.35;
        if (!best || savings > best.savings) best = { port: p, savings, shift };
      }
      if (best && best.savings > 0) {
        recs.push({
          ...base,
          option: "DIVERT",
          targetPort: best.port.name,
          etaShiftHrs: +best.shift.toFixed(1),
          estSavingsUsd: Math.round(best.savings),
          rationale: `Predicted ${wait.toFixed(0)}h wait at destination vs ${best.port.transitHrsAt15kn}h transit to ${best.port.name} (${best.port.availability} availability). Diversion avoids ~${Math.max(0, wait - best.shift).toFixed(0)}h of anchorage time; net saving ≈ $${Math.round(best.savings).toLocaleString()} at $32k/day ship cost.`,
          confidence: 0.82,
          tier: "critical",
        });
        continue;
      }
    }

    // ------------------------------------------------------------ SLOW_STEAM
    if (wait >= 18 && wait < 48 && v.status === "INBOUND") {
      // vessel can cut speed to arrive as the window opens; ~35% fuel-burn
      // reduction steaming at ~60% power (documented slow-steaming figure)
      const steamDownHrs = Math.min(wait - 6, 48);
      const savings = 0.35 * (steamDownHrs / 24) * DAILY_OP_COST_USD;
      recs.push({
        ...base,
        option: "SLOW_STEAM",
        etaShiftHrs: +steamDownHrs.toFixed(1),
        estSavingsUsd: Math.round(savings),
        rationale: `Inbound with predicted ${wait.toFixed(0)}h queue. Reduce to ~14 kn so arrival coincides with the freed window; ~35% fuel-burn reduction over ${steamDownHrs.toFixed(0)}h saves ≈ $${Math.round(savings).toLocaleString()} and frees anchorage space.`,
        confidence: 0.72,
        tier: "high",
      });
      continue;
    }

    // -------------------------------------------------------- PRIORITY_WINDOW
    if (wait >= 10 && v.reeferUnits >= 200) {
      const savings = v.reeferUnits * REEFER_CONTENT_VALUE_USD * 0.4;
      recs.push({
        ...base,
        option: "PRIORITY_WINDOW",
        etaShiftHrs: -Math.min(6, wait - 8),
        estSavingsUsd: Math.round(savings),
        rationale: `${v.reeferUnits} reefer units aboard with predicted ${wait.toFixed(0)}h wait. Request priority window swap with a lower-priority call; avoids spoilage risk ≈ $${Math.round(savings).toLocaleString()} and reduces genset load at anchor.`,
        confidence: 0.64,
        tier: "medium",
      });
      continue;
    }

    // ------------------------------------------------------------------ HOLD
    recs.push({
      ...base,
      option: "HOLD",
      etaShiftHrs: 0,
      estSavingsUsd: 0,
      rationale:
        wait < 10
          ? `Predicted ${wait.toFixed(0)}h wait is within normal rotation — hold current schedule.`
          : `Moderate ${wait.toFixed(0)}h wait but no reefer/size constraint and diversion economics negative — hold and re-evaluate next cycle.`,
      confidence: 0.55,
      tier: "low",
    });
  }

  return recs.sort(
    (a, b) =>
      ({ critical: 0, high: 1, medium: 2, low: 3 })[a.tier] -
      ({ critical: 0, high: 1, medium: 2, low: 3 })[b.tier] ||
      b.estSavingsUsd - a.estSavingsUsd,
  );
}
